File: a-stock-daily/scripts/eastmoney_adjusted_bars.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import os
import sys
import time
import urllib.parse
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed


def env(name: str, default: str) -> str:
    return os.environ.get(name, default).strip() or default


def clickhouse_endpoint() -> str:
    params = urllib.parse.urlencode(
        {
            "user": env("CLICKHOUSE_USER", "stock_user"),
            "password": env("CLICKHOUSE_PASSWORD", "stock_pass"),
            "database": env("CLICKHOUSE_DATABASE", "stock_db"),
        }
    )
    return f"{env('CLICKHOUSE_URL', 'http://a-stock-clickhouse:8123')}/?{params}"


def clickhouse_query(sql: str) -> str:
    request = urllib.request.Request(clickhouse_endpoint(), data=sql.encode("utf-8"), method="POST")
    with urllib.request.urlopen(request, timeout=int(env("CLICKHOUSE_TIMEOUT", "60"))) as response:
        return response.read().decode("utf-8")


def load_symbols() -> list[str]:
    text = clickhouse_query("SELECT code FROM stock_info ORDER BY code FORMAT JSONEachRow")
    return [json.loads(line)["code"] for line in text.splitlines() if line.strip()]


def secid(symbol: str) -> str:
    market = "1" if symbol.startswith(("6", "5")) else "0"
    return f"{market}.{symbol}"


def fqt(adjust: str) -> str:
    if adjust == "qfq":
        return "1"
    if adjust == "hfq":
        return "2"
    return "0"


def fetch_symbol(symbol: str, adjust: str, start_date: str, end_date: str) -> list[dict[str, object]]:
    params = urllib.parse.urlencode(
        {
            "fields1": "f1,f2,f3,f4,f5,f6",
            "fields2": "f51,f52,f53,f54,f55,f56,f57,f58,f59,f60,f61",
            "klt": "101",
            "fqt": fqt(adjust),
            "beg": start_date,
            "end": end_date,
            "secid": secid(symbol),
        }
    )
    url = f"https://push2his.eastmoney.com/api/qt/stock/kline/get?{params}"
    request = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
    with urllib.request.urlopen(request, timeout=int(env("EASTMONEY_TIMEOUT_SECONDS", "30"))) as response:
        payload = json.loads(response.read().decode("utf-8"))

    klines = ((payload.get("data") or {}).get("klines") or [])
    rows: list[dict[str, object]] = []
    for line in klines:
        parts = str(line).split(",")
        if len(parts) < 7:
            continue
        trade_date = parts[0]
        rows.append(
            {
                "code": symbol,
                "adjust": adjust,
                "trade_date": trade_date,
                "datetime": f"{trade_date} 15:00:00",
                "date_str": trade_date,
                "open": float(parts[1] or 0),
                "close": float(parts[2] or 0),
                "high": float(parts[3] or 0),
                "low": float(parts[4] or 0),
                "volume": int(float(parts[5] or 0)),
                "amount": float(parts[6] or 0),
                "source": "eastmoney",
            }
        )
    return rows


def ensure_table() -> None:
    clickhouse_query(
        """
CREATE TABLE IF NOT EXISTS tdx_daily_bars
(
    code String,
    adjust String DEFAULT 'none',
    trade_date Date,
    datetime String,
    date_str String,
    open Float64,
    close Float64,
    high Float64,
    low Float64,
    volume UInt64,
    amount Float64,
    source String DEFAULT 'tdx',
    updated_at DateTime DEFAULT now()
)
ENGINE = ReplacingMergeTree(updated_at)
ORDER BY (code, adjust, trade_date)
"""
    )


def insert_rows(rows: list[dict[str, object]]) -> None:
    if not rows:
        return
    payload = "\n".join(json.dumps(row, ensure_ascii=False, separators=(",", ":")) for row in rows)
    clickhouse_query(f"INSERT INTO tdx_daily_bars FORMAT JSONEachRow\n{payload}\n")


def run_symbol(symbol: str, adjusts: list[str], start_date: str, end_date: str) -> tuple[str, int, str]:
    try:
        total = 0
        for adjust in adjusts:
            rows = fetch_symbol(symbol, adjust, start_date, end_date)
            insert_rows(rows)
            total += len(rows)
        return symbol, 0, f"inserted {total}"
    except Exception as exc:
        return symbol, 1, str(exc)


def main() -> int:
    mode = sys.argv[1] if len(sys.argv) > 1 else "history"
    start_date = (sys.argv[2] if len(sys.argv) > 2 else env("TDX_START_DATE", "2010-01-01")).replace("-", "")
    end_date = (sys.argv[3] if len(sys.argv) > 3 else env("TDX_END_DATE", "20500101")).replace("-", "")
    adjusts = [item.strip() for item in env("ADJUSTED_BARS_ADJUSTS", "qfq,hfq").split(",") if item.strip()]
    workers = int(env("ADJUSTED_BARS_WORKERS", env("TDX_BACKFILL_WORKERS", "8")))
    retries = int(env("TDX_RETRIES", "3"))

    ensure_table()
    symbols = load_symbols()
    if not symbols:
        raise RuntimeError("no symbols found in stock_info")

    print(f"eastmoney {mode} start={start_date} end={end_date} adjusts={','.join(adjusts)} symbols={len(symbols)} workers={workers}", flush=True)
    pending = symbols
    total_ok = 0
    failed: list[tuple[str, str]] = []
    for attempt in range(retries + 1):
        failed = []
        started = time.time()
        with ThreadPoolExecutor(max_workers=workers) as pool:
            futures = [pool.submit(run_symbol, symbol, adjusts, start_date, end_date) for symbol in pending]
            for index, future in enumerate(as_completed(futures), start=1):
                symbol, code, output = future.result()
                if code == 0:
                    total_ok += 1
                else:
                    failed.append((symbol, output))
                    print(f"failed {symbol}: {output}", file=sys.stderr, flush=True)
                if index % 50 == 0 or index == len(pending):
                    elapsed = max(0.1, time.time() - started)
                    print(f"progress {index}/{len(pending)} ok={index-len(failed)} failed={len(failed)} speed={index/elapsed:.2f}/s", flush=True)
        if not failed:
            print(f"eastmoney {mode} finished ok={total_ok}", flush=True)
            return 0
        pending = [symbol for symbol, _ in failed]
        print(f"retry attempt {attempt + 1}/{retries} symbols={len(pending)}", flush=True)
        time.sleep(float(env("TDX_RETRY_SLEEP_SECONDS", "5")))

    print(f"eastmoney {mode} finished with failures: {len(failed)}", file=sys.stderr)
    return 1


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except Exception as exc:
        print(f"eastmoney adjusted bars failed: {exc}", file=sys.stderr)
        raise SystemExit(1)

File: a-stock-daily/src/bin/eastmoney_adjusted_bars.rs
use anyhow::{anyhow, Context, Result};
use chrono::{Datelike, Local};
use serde::{Deserialize, Serialize};
use std::env;
use std::io::{Read, Write};
use std::net::TcpStream;
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

#[derive(Debug, Clone, Serialize)]
struct DailyBar {
    code: String,
    adjust: String,
    trade_date: String,
    datetime: String,
    date_str: String,
    open: f64,
    close: f64,
    high: f64,
    low: f64,
    volume: u64,
    amount: f64,
    source: String,
}

#[derive(Debug, Deserialize)]
struct EastmoneyResponse {
    rc: i64,
    data: Option<EastmoneyData>,
}

#[derive(Debug, Deserialize)]
struct EastmoneyData {
    klines: Option<Vec<String>>,
}

fn main() -> Result<()> {
    let args: Vec<String> = env::args().collect();
    let mode = args.get(1).map(String::as_str).unwrap_or("history");
    let default_start;
    let start_raw = if let Some(value) = args.get(2) {
        value.as_str()
    } else {
        default_start = if mode == "daily" {
            today_ymd_string()
        } else {
            env_string("TDX_START_DATE", "2010-01-01")
        };
        default_start.as_str()
    };
    let start_date = normalize_date(start_raw)?;
    let default_end;
    let end_raw = if let Some(value) = args.get(3) {
        value.as_str()
    } else {
        default_end = if mode == "daily" {
            today_ymd_string()
        } else {
            env_string("TDX_END_DATE", "20500101")
        };
        default_end.as_str()
    };
    let end_date = normalize_date(end_raw)?;
    let adjusts = env_string("ADJUSTED_BARS_ADJUSTS", "qfq,hfq")
        .split(',')
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_string)
        .collect::<Vec<_>>();
    let workers = env_usize("ADJUSTED_BARS_WORKERS", env_usize("TDX_BACKFILL_WORKERS", 8)? as u64)?;
    let retries = env_usize("TDX_RETRIES", 3)?;

    ensure_table()?;
    let symbols = load_symbols()?;
    if symbols.is_empty() {
        return Err(anyhow!("no symbols found in stock_info"));
    }

    println!(
        "eastmoney {mode} start={start_date} end={end_date} adjusts={} symbols={} workers={workers}",
        adjusts.join(","),
        symbols.len()
    );
    let mut pending = symbols;
    let mut total_ok = 0usize;
    let mut failed = Vec::new();

    for attempt in 0..=retries {
        if attempt > 0 {
            println!("retry attempt {attempt}/{retries} symbols={}", pending.len());
            thread::sleep(Duration::from_secs(env_u64("TDX_RETRY_SLEEP_SECONDS", 5)?));
        }
        let result = run_batch(&pending, &adjusts, &start_date, &end_date, workers)?;
        total_ok += result.ok;
        failed = result.failed;
        if failed.is_empty() {
            println!("eastmoney {mode} finished ok={total_ok}");
            return Ok(());
        }
        pending = failed.iter().map(|(symbol, _)| symbol.clone()).collect();
    }

    eprintln!("eastmoney {mode} finished with failures: {}", failed.len());
    Err(anyhow!("failed symbols remain: {}", failed.len()))
}

struct BatchResult {
    ok: usize,
    failed: Vec<(String, String)>,
}

fn run_batch(symbols: &[String], adjusts: &[String], start_date: &str, end_date: &str, workers: usize) -> Result<BatchResult> {
    let queue = Arc::new(Mutex::new(symbols.to_vec()));
    let ok = Arc::new(Mutex::new(0usize));
    let done = Arc::new(Mutex::new(0usize));
    let failed = Arc::new(Mutex::new(Vec::<(String, String)>::new()));
    let total = symbols.len();

    let mut handles = Vec::new();
    for _ in 0..workers.max(1) {
        let queue = Arc::clone(&queue);
        let ok = Arc::clone(&ok);
        let done = Arc::clone(&done);
        let failed = Arc::clone(&failed);
        let adjusts = adjusts.to_vec();
        let start_date = start_date.to_string();
        let end_date = end_date.to_string();
        handles.push(thread::spawn(move || loop {
            let symbol = {
                let mut queue = queue.lock().expect("queue lock");
                queue.pop()
            };
            let Some(symbol) = symbol else {
                break;
            };

            match run_symbol(&symbol, &adjusts, &start_date, &end_date) {
                Ok(()) => *ok.lock().expect("ok lock") += 1,
                Err(exc) => {
                    eprintln!("failed {symbol}: {exc:#}");
                    failed.lock().expect("failed lock").push((symbol.clone(), format!("{exc:#}")));
                }
            }

            let current_done = {
                let mut done = done.lock().expect("done lock");
                *done += 1;
                *done
            };
            if current_done % 50 == 0 || current_done == total {
                let current_ok = *ok.lock().expect("ok lock");
                let current_failed = failed.lock().expect("failed lock").len();
                println!("progress {current_done}/{total} ok={current_ok} failed={current_failed}");
            }
        }));
    }

    for handle in handles {
        handle.join().map_err(|_| anyhow!("worker thread panicked"))?;
    }

    let ok_count = *ok.lock().expect("ok lock");
    let failed_items = failed.lock().expect("failed lock").clone();
    Ok(BatchResult {
        ok: ok_count,
        failed: failed_items,
    })
}

fn run_symbol(symbol: &str, adjusts: &[String], start_date: &str, end_date: &str) -> Result<()> {
    let mut total = 0usize;
    for adjust in adjusts {
        let rows = fetch_symbol(symbol, adjust, start_date, end_date)?;
        total += rows.len();
        insert_rows(&rows)?;
    }
    eprintln!("inserted {total} rows for {symbol}");
    Ok(())
}

fn fetch_symbol(symbol: &str, adjust: &str, start_date: &str, end_date: &str) -> Result<Vec<DailyBar>> {
    let url = format!(
        "https://push2his.eastmoney.com/api/qt/stock/kline/get?fields1=f1%2Cf2%2Cf3%2Cf4%2Cf5%2Cf6&fields2=f51%2Cf52%2Cf53%2Cf54%2Cf55%2Cf56%2Cf57%2Cf58%2Cf59%2Cf60%2Cf61&klt=101&fqt={}&beg={}&end={}&secid={}",
        fqt(adjust)?,
        start_date,
        end_date,
        secid(symbol)
    );
    let body = ureq::get(&url)
        .set("User-Agent", "Mozilla/5.0")
        .timeout(Duration::from_secs(env_u64("EASTMONEY_TIMEOUT_SECONDS", 30)?))
        .call()
        .with_context(|| format!("eastmoney request failed for {symbol} {adjust}"))?
        .into_string()?;
    let response: EastmoneyResponse = serde_json::from_str(&body)?;
    if response.rc != 0 {
        return Err(anyhow!("eastmoney rc={} for {symbol} {adjust}", response.rc));
    }

    let mut rows = Vec::new();
    for line in response.data.and_then(|data| data.klines).unwrap_or_default() {
        let parts = line.split(',').collect::<Vec<_>>();
        if parts.len() < 7 {
            continue;
        }
        let trade_date = parts[0].to_string();
        rows.push(DailyBar {
            code: symbol.to_string(),
            adjust: adjust.to_string(),
            trade_date: trade_date.clone(),
            datetime: format!("{trade_date} 15:00:00"),
            date_str: trade_date,
            open: parse_f64(parts[1])?,
            close: parse_f64(parts[2])?,
            high: parse_f64(parts[3])?,
            low: parse_f64(parts[4])?,
            volume: parse_f64(parts[5])?.round() as u64,
            amount: parse_f64(parts[6])?,
            source: "eastmoney".to_string(),
        });
    }
    Ok(rows)
}

fn ensure_table() -> Result<()> {
    clickhouse_post(
        r#"
CREATE TABLE IF NOT EXISTS tdx_daily_bars
(
    code String,
    adjust String DEFAULT 'none',
    trade_date Date,
    datetime String,
    date_str String,
    open Float64,
    close Float64,
    high Float64,
    low Float64,
    volume UInt64,
    amount Float64,
    source String DEFAULT 'tdx',
    updated_at DateTime DEFAULT now()
)
ENGINE = ReplacingMergeTree(updated_at)
ORDER BY (code, adjust, trade_date)
"#,
    )?;
    Ok(())
}

fn load_symbols() -> Result<Vec<String>> {
    let text = clickhouse_post("SELECT code FROM stock_info ORDER BY code FORMAT JSONEachRow")?;
    let mut symbols = Vec::new();
    for line in text.lines().filter(|line| !line.trim().is_empty()) {
        let value: serde_json::Value = serde_json::from_str(line)?;
        if let Some(code) = value.get("code").and_then(|code| code.as_str()) {
            symbols.push(code.to_string());
        }
    }
    Ok(symbols)
}

fn insert_rows(rows: &[DailyBar]) -> Result<()> {
    if rows.is_empty() {
        return Ok(());
    }
    let mut body = String::from("INSERT INTO tdx_daily_bars FORMAT JSONEachRow\n");
    for row in rows {
        body.push_str(&serde_json::to_string(row)?);
        body.push('\n');
    }
    clickhouse_post(&body)?;
    Ok(())
}

fn clickhouse_post(sql: &str) -> Result<String> {
    let endpoint = env_string("CLICKHOUSE_URL", "http://a-stock-clickhouse:8123");
    let endpoint = endpoint.trim_end_matches('/').strip_prefix("http://").ok_or_else(|| anyhow!("only http clickhouse url is supported"))?;
    let (authority, path) = endpoint.split_once('/').unwrap_or((endpoint, ""));
    let (host, port) = if let Some((host, port)) = authority.rsplit_once(':') {
        (host.to_string(), port.parse::<u16>()?)
    } else {
        (authority.to_string(), 80)
    };
    let path = if path.is_empty() { "/".to_string() } else { format!("/{path}") };
    let query = format!(
        "database={}&user={}&password={}",
        url_component(&env_string("CLICKHOUSE_DATABASE", "stock_db")),
        url_component(&env_string("CLICKHOUSE_USER", "stock_user")),
        url_component(&env_string("CLICKHOUSE_PASSWORD", "stock_pass"))
    );
    let full_path = format!("{path}?{query}");
    let request = format!(
        "POST {full_path} HTTP/1.1\r\nHost: {authority}\r\nContent-Type: text/plain\r\nContent-Length: {}\r\nConnection: close\r\n\r\n",
        sql.len()
    );
    let mut stream = TcpStream::connect((host.as_str(), port))?;
    stream.write_all(request.as_bytes())?;
    stream.write_all(sql.as_bytes())?;
    stream.shutdown(std::net::Shutdown::Write)?;
    let mut response = Vec::new();
    stream.read_to_end(&mut response)?;
    let (status, body) = parse_http_response(&response)?;
    if !(200..300).contains(&status) {
        return Err(anyhow!("ClickHouse returned HTTP {status}: {body}"));
    }
    Ok(body)
}

fn parse_http_response(response: &[u8]) -> Result<(u16, String)> {
    let header_end = response.windows(4).position(|window| window == b"\r\n\r\n").ok_or_else(|| anyhow!("invalid HTTP response"))?;
    let header_text = String::from_utf8_lossy(&response[..header_end]);
    let status = header_text.lines().next().and_then(|line| line.split_whitespace().nth(1)).ok_or_else(|| anyhow!("missing HTTP status"))?.parse::<u16>()?;
    let raw_body = &response[(header_end + 4)..];
    let is_chunked = header_text
        .lines()
        .any(|line| line.to_ascii_lowercase().starts_with("transfer-encoding:") && line.to_ascii_lowercase().contains("chunked"));
    let body_bytes = if is_chunked {
        decode_chunked(raw_body)?
    } else {
        raw_body.to_vec()
    };
    let body = String::from_utf8_lossy(&body_bytes).to_string();
    Ok((status, body))
}

fn decode_chunked(mut body: &[u8]) -> Result<Vec<u8>> {
    let mut out = Vec::new();
    loop {
        let line_end = body
            .windows(2)
            .position(|window| window == b"\r\n")
            .ok_or_else(|| anyhow!("invalid chunked response: missing size line"))?;
        let size_text = String::from_utf8_lossy(&body[..line_end]);
        let size_hex = size_text.split(';').next().unwrap_or("").trim();
        let size = usize::from_str_radix(size_hex, 16)
            .with_context(|| format!("invalid chunk size: {size_text}"))?;
        body = &body[(line_end + 2)..];
        if size == 0 {
            break;
        }
        if body.len() < size + 2 {
            return Err(anyhow!("invalid chunked response: incomplete chunk"));
        }
        out.extend_from_slice(&body[..size]);
        body = &body[(size + 2)..];
    }
    Ok(out)
}

fn fqt(adjust: &str) -> Result<&'static str> {
    match adjust {
        "qfq" | "forward" | "forward_additive" => Ok("1"),
        "hfq" | "backward" | "backward_additive" => Ok("2"),
        "none" | "raw" => Ok("0"),
        other => Err(anyhow!("unsupported adjust: {other}; expected qfq|hfq|none")),
    }
}

fn secid(symbol: &str) -> String {
    let market = if symbol.starts_with('6') || symbol.starts_with('5') { "1" } else { "0" };
    format!("{market}.{symbol}")
}

fn normalize_date(raw: &str) -> Result<String> {
    let value = raw.replace('-', "");
    if value.len() != 8 || !value.chars().all(|ch| ch.is_ascii_digit()) {
        return Err(anyhow!("invalid date: {raw}"));
    }
    Ok(value)
}

fn today_ymd_string() -> String {
    let today = Local::now();
    format!("{:04}{:02}{:02}", today.year(), today.month(), today.day())
}

fn parse_f64(value: &str) -> Result<f64> {
    Ok(value.parse::<f64>().unwrap_or(0.0))
}

fn url_component(value: &str) -> String {
    value
        .bytes()
        .flat_map(|byte| match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => vec![byte as char],
            _ => format!("%{byte:02X}").chars().collect(),
        })
        .collect()
}

fn env_string(name: &str, default: &str) -> String {
    env::var(name).ok().map(|value| value.trim().to_string()).filter(|value| !value.is_empty()).unwrap_or_else(|| default.to_string())
}

fn env_u64(name: &str, default: u64) -> Result<u64> {
    match env::var(name) {
        Ok(value) if !value.trim().is_empty() => Ok(value.parse()?),
        _ => Ok(default),
    }
}

fn env_usize(name: &str, default: u64) -> Result<usize> {
    Ok(env_u64(name, default)? as usize)
}
